commit 8fcc2808ccb474b8740af16524f5e1d68ae7fca8
Author: Milad Kakoli <hellcrueldictator@yahoo.com>
Date:   Fri Jun 1 11:42:35 2012 +0430

    add persian keyboard layout from gingerbread
    
    persian keyboard layout on gingerbread isn't correct,and in ics there
    is'nt any,this is the correct one,

diff --git a/java/res/xml-fa/debug.log b/java/res/xml-fa/debug.log
new file mode 100644
index 0000000..e69de29
diff --git a/java/res/xml-fa/kbd_qwerty.xml b/java/res/xml-fa/kbd_qwerty.xml
new file mode 100644
index 0000000..a318ca6
--- /dev/null
+++ b/java/res/xml-fa/kbd_qwerty.xml
@@ -0,0 +1,493 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+/*
+**
+** Copyright 2010, The Android Open Source Project
+**
+** Licensed under the Apache License, Version 2.0 (the "License");
+** you may not use this file except in compliance with the License.
+** You may obtain a copy of the License at
+**
+**     http://www.apache.org/licenses/LICENSE-2.0
+**
+** Unless required by applicable law or agreed to in writing, software
+** distributed under the License is distributed on an "AS IS" BASIS,
+** WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+** See the License for the specific language governing permissions and
+** limitations under the License.
+*/
+-->
+<!-- By Eyad Aboulouz -->
+<!-- Modified by Milad Kakoli -->
+<Keyboard
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    android:keyWidth="10%p"
+    android:horizontalGap="0px"
+    android:verticalGap="@dimen/key_bottom_gap"
+    android:keyHeight="@dimen/key_height"
+>
+    <Row>
+        <Key
+            android:keyLabel="╪╢"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="┘┘ï┘┘î┘┘┘ّ┘ْ"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:keyLabel="╪╡" />
+        <Key
+            android:keyLabel="┘é" />
+        <Key
+            android:keyLabel="┘" />
+        <Key
+            android:keyLabel="╪║" />
+        <Key
+            android:keyLabel="╪╣" />
+        <Key
+            android:keyLabel="┘ç"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="╪ر" />
+        <Key
+            android:keyLabel="╪«" />
+        <Key
+            android:keyLabel="╪ص" />
+        <Key
+            android:keyLabel="╪ش"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="┌"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row>
+        <Key
+            android:keyLabel="╪┤"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:keyLabel="╪│" />
+        <Key
+            android:keyLabel="┘è"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="┘ë╪خ" />
+        <Key
+            android:keyLabel="╪ذ"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="┘╛" />
+        <Key
+            android:keyLabel="┘"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="ي╗╡ي╗╣ي╗╖ي╗╗" />
+        <Key
+            android:keyLabel="╪د"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="╪ث╪ح╪ت╪ة" />
+        <Key
+            android:keyLabel="╪ز" />
+        <Key
+            android:keyLabel="┘" />
+        <Key
+            android:keyLabel="┘à" />
+        <Key
+            android:keyLabel="┌ر"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="┌»"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row>
+        <Key
+            android:keyLabel="╪╕"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="╪ا╪î╪ؤ"
+            android:horizontalGap="2.5%p"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:keyLabel="╪╖" />
+        <Key
+            android:keyLabel="╪▓"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="┌ء"  />
+        <Key
+            android:keyLabel="╪▒" />
+		<Key
+            android:keyLabel="╪░" />
+        <Key
+            android:keyLabel="╪»" />
+        <Key
+            android:keyLabel="┘ê"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="╪ج"  />
+        <Key
+            android:keyLabel="╪س" />
+        <Key
+            android:codes="@integer/key_delete"
+            android:keyIcon="@drawable/sym_keyboard_delete"
+            android:iconPreview="@drawable/sym_keyboard_feedback_delete"
+            android:keyWidth="15%p"
+            android:isRepeatable="true"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row
+        android:keyboardMode="@+id/mode_normal"
+        android:keyWidth="10%p"
+        android:rowEdgeFlags="bottom"
+    >
+        <Key
+            android:codes="@integer/key_symbol"
+            android:keyLabel="@string/label_symbol_key"
+            android:keyWidth="20%p"
+            android:isModifier="true"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:codes="@integer/key_f1"
+            android:isModifier="true" />
+        <Key
+            android:codes="@integer/key_space"
+            android:keyIcon="@drawable/sym_keyboard_space"
+            android:iconPreview="@drawable/sym_keyboard_feedback_space"
+            android:keyWidth="40%p"
+            android:isModifier="true" />
+        <Key
+            android:keyLabel="."
+            android:keyIcon="@drawable/hint_popup"
+            android:popupKeyboard="@xml/popup_punctuation"
+            android:isModifier="true" />
+        <Key
+            android:codes="@integer/key_return"
+            android:keyIcon="@drawable/sym_keyboard_return"
+            android:iconPreview="@drawable/sym_keyboard_feedback_return"
+            android:keyWidth="20%p"
+            android:isModifier="true"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row
+        android:keyboardMode="@+id/mode_url"
+        android:keyWidth="10%p"
+        android:rowEdgeFlags="bottom"
+    >
+        <Key
+            android:codes="@integer/key_symbol"
+            android:keyLabel="@string/label_symbol_key"
+            android:keyWidth="20%p"
+            android:isModifier="true"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:codes="@integer/key_f1"
+            android:isModifier="true" />
+        <Key
+            android:codes="@integer/key_space"
+            android:keyIcon="@drawable/sym_keyboard_space"
+            android:iconPreview="@drawable/sym_keyboard_feedback_space"
+            android:keyWidth="40%p"
+            android:isModifier="true" />
+        <Key
+            android:keyLabel="."
+            android:keyIcon="@drawable/hint_popup"
+            android:popupKeyboard="@xml/popup_punctuation"
+            android:isModifier="true" />
+        <Key
+            android:codes="@integer/key_return"
+            android:keyIcon="@drawable/sym_keyboard_return"
+            android:iconPreview="@drawable/sym_keyboard_feedback_return"
+            android:keyWidth="20%p"
+            android:isModifier="true"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row
+        android:keyboardMode="@+id/mode_email"
+        android:keyWidth="10%p"
+        android:rowEdgeFlags="bottom"
+    >
+        <Key
+            android:codes="@integer/key_symbol"
+            android:keyLabel="@string/label_symbol_key"
+            android:keyWidth="20%p"
+            android:isModifier="true"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:codes="@integer/key_f1"
+            android:isModifier="true" />
+        <Key
+            android:codes="@integer/key_space"
+            android:keyIcon="@drawable/sym_keyboard_space"
+            android:iconPreview="@drawable/sym_keyboard_feedback_space"
+            android:keyWidth="40%p"
+            android:isModifier="true" />
+        <Key
+            android:keyLabel="."
+            android:keyIcon="@drawable/hint_popup"
+            android:popupKeyboard="@xml/popup_punctuation"
+            android:isModifier="true" />
+        <Key
+            android:codes="@integer/key_return"
+            android:keyIcon="@drawable/sym_keyboard_return"
+            android:iconPreview="@drawable/sym_keyboard_feedback_return"
+            android:keyWidth="20%p"
+            android:isModifier="true"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row
+        android:keyboardMode="@+id/mode_im"
+        android:keyWidth="10%p"
+        android:rowEdgeFlags="bottom"
+    >
+        <Key
+            android:codes="@integer/key_symbol"
+            android:keyLabel="@string/label_symbol_key"
+            android:keyWidth="20%p"
+            android:isModifier="true"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:codes="@integer/key_f1"
+            android:isModifier="true" />
+        <Key
+            android:codes="@integer/key_space"
+            android:keyIcon="@drawable/sym_keyboard_space"
+            android:iconPreview="@drawable/sym_keyboard_feedback_space"
+            android:keyWidth="40%p"
+            android:isModifier="true" />
+        <Key
+            android:keyLabel="."
+            android:keyIcon="@drawable/hint_popup"
+            android:popupKeyboard="@xml/popup_punctuation"
+            android:isModifier="true" />
+        <Key
+            android:keyLabel="(-:"
+            android:keyOutputText=":-) "
+            android:keyIcon="@drawable/hint_popup"
+            android:popupKeyboard="@xml/popup_smileys"
+            android:keyWidth="20%p"
+            android:isModifier="true"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row
+        android:keyboardMode="@+id/mode_webentry"
+        android:keyWidth="10%p"
+        android:rowEdgeFlags="bottom"
+    >
+        <Key
+            android:codes="@integer/key_symbol"
+            android:keyLabel="@string/label_symbol_key"
+            android:keyWidth="20%p"
+            android:isModifier="true"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:codes="@integer/key_f1"
+            android:isModifier="true" />
+        <Key
+            android:codes="@integer/key_space"
+            android:keyIcon="@drawable/sym_keyboard_space"
+            android:iconPreview="@drawable/sym_keyboard_feedback_space"
+            android:keyWidth="30%p"
+            android:isModifier="true" />
+        <Key
+            android:codes="@integer/key_tab"
+            android:keyIcon="@drawable/sym_keyboard_tab"
+            android:iconPreview="@drawable/sym_keyboard_feedback_tab"
+            android:isModifier="true" />
+        <Key
+            android:keyLabel="."
+            android:keyIcon="@drawable/hint_popup"
+            android:popupKeyboard="@xml/popup_punctuation"
+            android:isModifier="true" />
+        <Key
+            android:codes="@integer/key_return"
+            android:keyIcon="@drawable/sym_keyboard_return"
+            android:iconPreview="@drawable/sym_keyboard_feedback_return"
+            android:keyWidth="20%p"
+            android:isModifier="true"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row
+        android:keyboardMode="@+id/mode_normal_with_settings_key"
+        android:keyWidth="10%p"
+        android:rowEdgeFlags="bottom"
+    >
+        <Key
+            android:codes="@integer/key_symbol"
+            android:keyLabel="@string/label_symbol_key"
+            android:keyWidth="15%p"
+            android:isModifier="true"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:codes="@integer/key_settings"
+            android:keyIcon="@drawable/sym_keyboard_settings"
+            android:iconPreview="@drawable/sym_keyboard_feedback_settings"
+            android:isModifier="true" />
+        <Key
+            android:codes="@integer/key_f1"
+            android:isModifier="true" />
+        <Key
+            android:codes="@integer/key_space"
+            android:keyIcon="@drawable/sym_keyboard_space"
+            android:iconPreview="@drawable/sym_keyboard_feedback_space"
+            android:keyWidth="30%p"
+            android:isModifier="true" />
+        <Key
+            android:keyLabel="."
+            android:keyIcon="@drawable/hint_popup"
+            android:popupKeyboard="@xml/popup_punctuation"
+            android:isModifier="true" />
+        <Key
+            android:codes="@integer/key_return"
+            android:keyIcon="@drawable/sym_keyboard_return"
+            android:iconPreview="@drawable/sym_keyboard_feedback_return"
+            android:keyWidth="25%p"
+            android:isModifier="true"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row
+        android:keyboardMode="@+id/mode_url_with_settings_key"
+        android:keyWidth="10%p"
+        android:rowEdgeFlags="bottom"
+    >
+        <Key
+            android:codes="@integer/key_symbol"
+            android:keyLabel="@string/label_symbol_key"
+            android:keyWidth="15%p"
+            android:isModifier="true"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:codes="@integer/key_settings"
+            android:keyIcon="@drawable/sym_keyboard_settings"
+            android:iconPreview="@drawable/sym_keyboard_feedback_settings"
+            android:isModifier="true" />
+        <Key
+            android:codes="@integer/key_f1"
+            android:isModifier="true" />
+        <Key
+            android:codes="@integer/key_space"
+            android:keyIcon="@drawable/sym_keyboard_space"
+            android:iconPreview="@drawable/sym_keyboard_feedback_space"
+            android:keyWidth="30%p"
+            android:isModifier="true" />
+        <Key
+            android:keyLabel="."
+            android:keyIcon="@drawable/hint_popup"
+            android:popupKeyboard="@xml/popup_punctuation"
+            android:isModifier="true" />
+        <Key
+            android:codes="@integer/key_return"
+            android:keyIcon="@drawable/sym_keyboard_return"
+            android:iconPreview="@drawable/sym_keyboard_feedback_return"
+            android:keyWidth="25%p"
+            android:isModifier="true"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row
+        android:keyboardMode="@+id/mode_email_with_settings_key"
+        android:keyWidth="10%p"
+        android:rowEdgeFlags="bottom"
+    >
+        <Key
+            android:codes="@integer/key_symbol"
+            android:keyLabel="@string/label_symbol_key"
+            android:keyWidth="15%p"
+            android:isModifier="true"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:codes="@integer/key_settings"
+            android:keyIcon="@drawable/sym_keyboard_settings"
+            android:iconPreview="@drawable/sym_keyboard_feedback_settings"
+            android:isModifier="true" />
+        <Key
+            android:codes="@integer/key_f1"
+            android:isModifier="true" />
+        <Key
+            android:codes="@integer/key_space"
+            android:keyIcon="@drawable/sym_keyboard_space"
+            android:iconPreview="@drawable/sym_keyboard_feedback_space"
+            android:keyWidth="30%p"
+            android:isModifier="true" />
+        <Key
+            android:keyLabel="."
+            android:keyIcon="@drawable/hint_popup"
+            android:popupKeyboard="@xml/popup_punctuation"
+            android:isModifier="true" />
+        <Key
+            android:codes="@integer/key_return"
+            android:keyIcon="@drawable/sym_keyboard_return"
+            android:iconPreview="@drawable/sym_keyboard_feedback_return"
+            android:keyWidth="25%p"
+            android:isModifier="true"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row
+        android:keyboardMode="@+id/mode_im_with_settings_key"
+        android:keyWidth="10%p"
+        android:rowEdgeFlags="bottom"
+    >
+        <Key
+            android:codes="@integer/key_symbol"
+            android:keyLabel="@string/label_symbol_key"
+            android:keyWidth="15%p"
+            android:isModifier="true"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:codes="@integer/key_settings"
+            android:keyIcon="@drawable/sym_keyboard_settings"
+            android:iconPreview="@drawable/sym_keyboard_feedback_settings"
+            android:isModifier="true" />
+        <Key
+            android:codes="@integer/key_f1"
+            android:isModifier="true" />
+        <Key
+            android:codes="@integer/key_space"
+            android:keyIcon="@drawable/sym_keyboard_space"
+            android:iconPreview="@drawable/sym_keyboard_feedback_space"
+            android:keyWidth="30%p"
+            android:isModifier="true" />
+        <Key
+            android:keyLabel="."
+            android:keyIcon="@drawable/hint_popup"
+            android:popupKeyboard="@xml/popup_punctuation"
+            android:isModifier="true" />
+        <Key
+            android:keyLabel="(-:"
+            android:keyOutputText=":-) "
+            android:keyIcon="@drawable/hint_popup"
+            android:popupKeyboard="@xml/popup_smileys"
+            android:keyWidth="25%p"
+            android:isModifier="true"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row
+        android:keyboardMode="@+id/mode_webentry_with_settings_key"
+        android:keyWidth="10%p"
+        android:rowEdgeFlags="bottom"
+    >
+        <Key
+            android:codes="@integer/key_symbol"
+            android:keyLabel="@string/label_symbol_key"
+            android:keyWidth="15%p"
+            android:isModifier="true"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:codes="@integer/key_settings"
+            android:keyIcon="@drawable/sym_keyboard_settings"
+            android:iconPreview="@drawable/sym_keyboard_feedback_settings"
+            android:isModifier="true" />
+        <Key
+            android:codes="@integer/key_f1"
+            android:isModifier="true" />
+        <Key
+            android:codes="@integer/key_space"
+            android:keyIcon="@drawable/sym_keyboard_space"
+            android:iconPreview="@drawable/sym_keyboard_feedback_space"
+            android:keyWidth="30%p"
+            android:isModifier="true" />
+        <Key
+            android:codes="@integer/key_tab"
+            android:keyIcon="@drawable/sym_keyboard_tab"
+            android:iconPreview="@drawable/sym_keyboard_feedback_tab"
+            android:isModifier="true" />
+        <Key
+            android:keyLabel="."
+            android:keyIcon="@drawable/hint_popup"
+            android:popupKeyboard="@xml/popup_punctuation"
+            android:isModifier="true" />
+        <Key
+            android:codes="@integer/key_return"
+            android:keyIcon="@drawable/sym_keyboard_return"
+            android:iconPreview="@drawable/sym_keyboard_feedback_return"
+            android:keyWidth="15%p"
+            android:isModifier="true"
+            android:keyEdgeFlags="right" />
+    </Row>
+</Keyboard>
\ No newline at end of file
diff --git a/java/res/xml-fa/kbd_qwerty_black.xml b/java/res/xml-fa/kbd_qwerty_black.xml
new file mode 100644
index 0000000..cfa932f
--- /dev/null
+++ b/java/res/xml-fa/kbd_qwerty_black.xml
@@ -0,0 +1,436 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+/*
+**
+** Copyright 2010, The Android Open Source Project
+**
+** Licensed under the Apache License, Version 2.0 (the "License");
+** you may not use this file except in compliance with the License.
+** You may obtain a copy of the License at
+**
+**     http://www.apache.org/licenses/LICENSE-2.0
+**
+** Unless required by applicable law or agreed to in writing, software
+** distributed under the License is distributed on an "AS IS" BASIS,
+** WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+** See the License for the specific language governing permissions and
+** limitations under the License.
+*/
+-->
+<!-- By Eyad Aboulouz -->
+<!-- Modified by Milad Kakoli -->
+<Keyboard
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    android:keyWidth="10%p"
+    android:horizontalGap="0px"
+    android:verticalGap="@dimen/key_bottom_gap"
+    android:keyHeight="@dimen/key_height"
+>
+    <Row>
+        <Key
+            android:keyLabel="╪╢"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="┘┘ï┘┘î┘┘┘ّ┘ْ"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:keyLabel="╪╡" />
+        <Key
+            android:keyLabel="┘é" />
+        <Key
+            android:keyLabel="┘" />
+        <Key
+            android:keyLabel="╪║" />
+        <Key
+            android:keyLabel="╪╣" />
+        <Key
+            android:keyLabel="┘ç"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="╪ر" />
+        <Key
+            android:keyLabel="╪«" />
+        <Key
+            android:keyLabel="╪ص" />
+        <Key
+            android:keyLabel="╪ش"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="┌"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row>
+        <Key
+            android:keyLabel="╪┤"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:keyLabel="╪│" />
+        <Key
+            android:keyLabel="┘è"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="┘ë╪خ" />
+        <Key
+            android:keyLabel="╪ذ"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="┘╛" />
+        <Key
+            android:keyLabel="┘"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="ي╗╡ي╗╣ي╗╖ي╗╗" />
+        <Key
+            android:keyLabel="╪د"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="╪ث╪ح╪ت╪ة" />
+        <Key
+            android:keyLabel="╪ز" />
+        <Key
+            android:keyLabel="┘" />
+        <Key
+            android:keyLabel="┘à" />
+        <Key
+            android:keyLabel="┌ر"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="┌»"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row>
+        <Key
+            android:keyLabel="╪╕"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="╪ا╪î╪ؤ"
+            android:horizontalGap="2.5%p"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:keyLabel="╪╖" />
+        <Key
+            android:keyLabel="╪▓"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="┌ء"  />
+        <Key
+            android:keyLabel="╪▒" />
+		<Key
+            android:keyLabel="╪░" />
+        <Key
+            android:keyLabel="╪»" />
+        <Key
+            android:keyLabel="┘ê"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="╪ج"  />
+        <Key
+            android:keyLabel="╪س" />
+        <Key
+            android:codes="@integer/key_delete"
+            android:keyIcon="@drawable/sym_bkeyboard_delete"
+            android:iconPreview="@drawable/sym_keyboard_feedback_delete"
+            android:keyWidth="15%p"
+            android:isRepeatable="true"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row
+        android:keyboardMode="@+id/mode_normal"
+        android:keyWidth="10%p"
+        android:rowEdgeFlags="bottom"
+    >
+        <Key
+            android:codes="@integer/key_symbol"
+            android:keyLabel="@string/label_symbol_key"
+            android:keyWidth="20%p"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:codes="@integer/key_f1" />
+        <Key
+            android:codes="@integer/key_space"
+            android:keyIcon="@drawable/sym_bkeyboard_space"
+            android:iconPreview="@drawable/sym_keyboard_feedback_space"
+            android:keyWidth="40%p" />
+        <Key
+            android:keyLabel="."
+            android:keyIcon="@drawable/hint_popup"
+            android:popupKeyboard="@xml/popup_punctuation" />
+        <Key
+            android:codes="@integer/key_return"
+            android:keyIcon="@drawable/sym_bkeyboard_return"
+            android:iconPreview="@drawable/sym_keyboard_feedback_return"
+            android:keyWidth="20%p"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row
+        android:keyboardMode="@+id/mode_url"
+        android:keyWidth="10%p"
+        android:rowEdgeFlags="bottom"
+    >
+        <Key
+            android:codes="@integer/key_symbol"
+            android:keyLabel="@string/label_symbol_key"
+            android:keyWidth="20%p"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:codes="@integer/key_f1" />
+        <Key
+            android:codes="@integer/key_space"
+            android:keyIcon="@drawable/sym_bkeyboard_space"
+            android:iconPreview="@drawable/sym_keyboard_feedback_space"
+            android:keyWidth="40%p" />
+        <Key
+            android:keyLabel="."
+            android:keyIcon="@drawable/hint_popup"
+            android:popupKeyboard="@xml/popup_punctuation" />
+        <Key
+            android:codes="@integer/key_return"
+            android:keyIcon="@drawable/sym_bkeyboard_return"
+            android:iconPreview="@drawable/sym_keyboard_feedback_return"
+            android:keyWidth="20%p"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row
+        android:keyboardMode="@+id/mode_email"
+        android:keyWidth="10%p"
+        android:rowEdgeFlags="bottom"
+    >
+        <Key
+            android:codes="@integer/key_symbol"
+            android:keyLabel="@string/label_symbol_key"
+            android:keyWidth="20%p"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:codes="@integer/key_f1" />
+        <Key
+            android:codes="@integer/key_space"
+            android:keyIcon="@drawable/sym_bkeyboard_space"
+            android:iconPreview="@drawable/sym_keyboard_feedback_space"
+            android:keyWidth="40%p" />
+        <Key
+            android:keyLabel="."
+            android:keyIcon="@drawable/hint_popup"
+            android:popupKeyboard="@xml/popup_punctuation" />
+        <Key
+            android:codes="@integer/key_return"
+            android:keyIcon="@drawable/sym_bkeyboard_return"
+            android:iconPreview="@drawable/sym_keyboard_feedback_return"
+            android:keyWidth="20%p"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row
+        android:keyboardMode="@+id/mode_im"
+        android:keyWidth="10%p"
+        android:rowEdgeFlags="bottom"
+    >
+        <Key
+            android:codes="@integer/key_symbol"
+            android:keyLabel="@string/label_symbol_key"
+            android:keyWidth="20%p"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:codes="@integer/key_f1" />
+        <Key
+            android:codes="@integer/key_space"
+            android:keyIcon="@drawable/sym_bkeyboard_space"
+            android:iconPreview="@drawable/sym_keyboard_feedback_space"
+            android:keyWidth="40%p" />
+        <Key
+            android:keyLabel="."
+            android:keyIcon="@drawable/hint_popup"
+            android:popupKeyboard="@xml/popup_punctuation" />
+        <Key
+            android:keyLabel="(-:"
+            android:keyOutputText=":-) "
+            android:keyIcon="@drawable/hint_popup"
+            android:popupKeyboard="@xml/popup_smileys"
+            android:keyWidth="20%p"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row
+        android:keyboardMode="@+id/mode_webentry"
+        android:keyWidth="10%p"
+        android:rowEdgeFlags="bottom"
+    >
+        <Key
+            android:codes="@integer/key_symbol"
+            android:keyLabel="@string/label_symbol_key"
+            android:keyWidth="20%p"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:codes="@integer/key_f1" />
+        <Key
+            android:codes="@integer/key_space"
+            android:keyIcon="@drawable/sym_bkeyboard_space"
+            android:iconPreview="@drawable/sym_keyboard_feedback_space"
+            android:keyWidth="30%p" />
+        <Key
+            android:codes="@integer/key_tab"
+            android:keyIcon="@drawable/sym_bkeyboard_tab"
+            android:iconPreview="@drawable/sym_keyboard_feedback_tab" />
+        <Key
+            android:keyLabel="."
+            android:keyIcon="@drawable/hint_popup"
+            android:popupKeyboard="@xml/popup_punctuation" />
+        <Key
+            android:codes="@integer/key_return"
+            android:keyIcon="@drawable/sym_bkeyboard_return"
+            android:iconPreview="@drawable/sym_keyboard_feedback_return"
+            android:keyWidth="20%p"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row
+        android:keyboardMode="@+id/mode_normal_with_settings_key"
+        android:keyWidth="10%p"
+        android:rowEdgeFlags="bottom"
+    >
+        <Key
+            android:codes="@integer/key_symbol"
+            android:keyLabel="@string/label_symbol_key"
+            android:keyWidth="15%p"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:codes="@integer/key_settings"
+            android:keyIcon="@drawable/sym_bkeyboard_settings"
+            android:iconPreview="@drawable/sym_keyboard_feedback_settings" />
+        <Key
+            android:codes="@integer/key_f1" />
+        <Key
+            android:codes="@integer/key_space"
+            android:keyIcon="@drawable/sym_bkeyboard_space"
+            android:iconPreview="@drawable/sym_keyboard_feedback_space"
+            android:keyWidth="30%p" />
+        <Key
+            android:keyLabel="."
+            android:keyIcon="@drawable/hint_popup"
+            android:popupKeyboard="@xml/popup_punctuation" />
+        <Key
+            android:codes="@integer/key_return"
+            android:keyIcon="@drawable/sym_bkeyboard_return"
+            android:iconPreview="@drawable/sym_keyboard_feedback_return"
+            android:keyWidth="25%p"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row
+        android:keyboardMode="@+id/mode_url_with_settings_key"
+        android:keyWidth="10%p"
+        android:rowEdgeFlags="bottom"
+    >
+        <Key
+            android:codes="@integer/key_symbol"
+            android:keyLabel="@string/label_symbol_key"
+            android:keyWidth="15%p"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:codes="@integer/key_settings"
+            android:keyIcon="@drawable/sym_bkeyboard_settings"
+            android:iconPreview="@drawable/sym_keyboard_feedback_settings" />
+        <Key
+            android:codes="@integer/key_f1" />
+        <Key
+            android:codes="@integer/key_space"
+            android:keyIcon="@drawable/sym_bkeyboard_space"
+            android:iconPreview="@drawable/sym_keyboard_feedback_space"
+            android:keyWidth="30%p" />
+        <Key
+            android:keyLabel="."
+            android:keyIcon="@drawable/hint_popup"
+            android:popupKeyboard="@xml/popup_punctuation" />
+        <Key
+            android:codes="@integer/key_return"
+            android:keyIcon="@drawable/sym_bkeyboard_return"
+            android:iconPreview="@drawable/sym_keyboard_feedback_return"
+            android:keyWidth="25%p"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row
+        android:keyboardMode="@+id/mode_email_with_settings_key"
+        android:keyWidth="10%p"
+        android:rowEdgeFlags="bottom"
+    >
+        <Key
+            android:codes="@integer/key_symbol"
+            android:keyLabel="@string/label_symbol_key"
+            android:keyWidth="15%p"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:codes="@integer/key_settings"
+            android:keyIcon="@drawable/sym_bkeyboard_settings"
+            android:iconPreview="@drawable/sym_keyboard_feedback_settings" />
+        <Key
+            android:codes="@integer/key_f1" />
+        <Key
+            android:codes="@integer/key_space"
+            android:keyIcon="@drawable/sym_bkeyboard_space"
+            android:iconPreview="@drawable/sym_keyboard_feedback_space"
+            android:keyWidth="30%p" />
+        <Key
+            android:keyLabel="."
+            android:keyIcon="@drawable/hint_popup"
+            android:popupKeyboard="@xml/popup_punctuation" />
+        <Key
+            android:codes="@integer/key_return"
+            android:keyIcon="@drawable/sym_bkeyboard_return"
+            android:iconPreview="@drawable/sym_keyboard_feedback_return"
+            android:keyWidth="25%p"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row
+        android:keyboardMode="@+id/mode_im_with_settings_key"
+        android:keyWidth="10%p"
+        android:rowEdgeFlags="bottom"
+    >
+        <Key
+            android:codes="@integer/key_symbol"
+            android:keyLabel="@string/label_symbol_key"
+            android:keyWidth="15%p"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:codes="@integer/key_settings"
+            android:keyIcon="@drawable/sym_bkeyboard_settings"
+            android:iconPreview="@drawable/sym_keyboard_feedback_settings" />
+        <Key
+            android:codes="@integer/key_f1" />
+        <Key
+            android:codes="@integer/key_space"
+            android:keyIcon="@drawable/sym_bkeyboard_space"
+            android:iconPreview="@drawable/sym_keyboard_feedback_space"
+            android:keyWidth="30%p" />
+        <Key
+            android:keyLabel="."
+            android:keyIcon="@drawable/hint_popup"
+            android:popupKeyboard="@xml/popup_punctuation" />
+        <Key
+            android:keyLabel="(-:"
+            android:keyOutputText=":-) "
+            android:keyIcon="@drawable/hint_popup"
+            android:popupKeyboard="@xml/popup_smileys"
+            android:keyWidth="25%p"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row
+        android:keyboardMode="@+id/mode_webentry_with_settings_key"
+        android:keyWidth="10%p"
+        android:rowEdgeFlags="bottom"
+    >
+        <Key
+            android:codes="@integer/key_symbol"
+            android:keyLabel="@string/label_symbol_key"
+            android:keyWidth="15%p"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:codes="@integer/key_settings"
+            android:keyIcon="@drawable/sym_bkeyboard_settings"
+            android:iconPreview="@drawable/sym_keyboard_feedback_settings" />
+        <Key
+            android:codes="@integer/key_f1" />
+        <Key
+            android:codes="@integer/key_space"
+            android:keyIcon="@drawable/sym_bkeyboard_space"
+            android:iconPreview="@drawable/sym_keyboard_feedback_space"
+            android:keyWidth="30%p" />
+        <Key
+            android:codes="@integer/key_tab"
+            android:keyIcon="@drawable/sym_bkeyboard_tab"
+            android:iconPreview="@drawable/sym_keyboard_feedback_tab" />
+        <Key
+            android:keyLabel="."
+            android:keyIcon="@drawable/hint_popup"
+            android:popupKeyboard="@xml/popup_punctuation" />
+        <Key
+            android:codes="@integer/key_return"
+            android:keyIcon="@drawable/sym_bkeyboard_return"
+            android:iconPreview="@drawable/sym_keyboard_feedback_return"
+            android:keyWidth="15%p"
+            android:keyEdgeFlags="right" />
+    </Row>
+</Keyboard>
diff --git a/java/res/xml-fa/kbd_symbols.xml b/java/res/xml-fa/kbd_symbols.xml
new file mode 100644
index 0000000..3fdd5a8
--- /dev/null
+++ b/java/res/xml-fa/kbd_symbols.xml
@@ -0,0 +1,222 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+/*
+**
+** Copyright 2008, The Android Open Source Project
+**
+** Licensed under the Apache License, Version 2.0 (the "License");
+** you may not use this file except in compliance with the License.
+** You may obtain a copy of the License at
+**
+**     http://www.apache.org/licenses/LICENSE-2.0
+**
+** Unless required by applicable law or agreed to in writing, software
+** distributed under the License is distributed on an "AS IS" BASIS,
+** WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+** See the License for the specific language governing permissions and
+** limitations under the License.
+*/
+-->
+<!-- By Eyad Aboulouz -->
+<Keyboard
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    android:keyWidth="10%p"
+    android:horizontalGap="0px"
+    android:verticalGap="@dimen/key_bottom_gap"
+    android:keyHeight="@dimen/key_height"
+>
+    <Row
+        android:rowEdgeFlags="top"
+    >
+        <Key
+            android:keyLabel="┘ة"
+            android:keyEdgeFlags="left"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="1" />
+        <Key
+            android:keyLabel="┘ت"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="2" />
+        <Key
+            android:keyLabel="┘ث"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="3" />
+        <Key
+            android:keyLabel="█┤"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="4" />
+        <Key
+            android:keyLabel="█╡"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="5" />
+        <Key
+            android:keyLabel="█╢"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="6" />
+        <Key
+            android:keyLabel="┘د"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="7" />
+        <Key
+            android:keyLabel="┘ذ"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="8" />
+        <Key
+            android:keyLabel="┘ر"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="9" />
+        <Key
+            android:keyLabel="┘ب"
+            android:keyEdgeFlags="right"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="0" />
+    </Row>
+    <Row>
+        <Key
+            android:keyLabel="\@"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:keyLabel="\#" />
+        <Key
+            android:keyLabel="$"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="┬ت┬ثظéش┬حظéثظéجظé▒" />
+        <Key
+            android:keyLabel="┘ز"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="%" />
+        <Key
+            android:keyLabel="\u0026" />
+        <Key
+            android:keyLabel="*"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="ظءà┘ص" />
+        <Key
+            android:keyLabel="-"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="_ظôظ¤" />
+        <Key
+            android:keyLabel="+"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="┬▒" />
+        <Key
+            android:keyLabel="("
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="[{\u003C" />
+        <Key
+            android:keyLabel=")"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="]}\u003E"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row>
+        <Key
+            android:codes="@integer/key_shift"
+            android:keyLabel="@string/label_alt_key"
+            android:keyWidth="15%p"
+            android:isModifier="true"
+            android:isSticky="true"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:keyLabel="!"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="┬ة" />
+        <Key
+            android:keyLabel="\u0022"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="ظ£ظإ┬س┬╗╦إ" />
+        <Key
+            android:keyLabel="\'"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="ظءظآ" />
+        <Key
+            android:keyLabel=":" />
+        <Key
+            android:keyLabel="╪ؤ"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters=";" />
+        <Key
+            android:keyLabel="/" />
+        <Key
+            android:keyLabel="╪ا"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="\?┬┐" />
+        <Key
+            android:codes="@integer/key_delete"
+            android:keyIcon="@drawable/sym_keyboard_delete"
+            android:iconPreview="@drawable/sym_keyboard_feedback_delete"
+            android:keyWidth="15%p"
+            android:isModifier="true"
+            android:isRepeatable="true"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row
+        android:keyboardMode="@+id/mode_symbols"
+        android:rowEdgeFlags="bottom"
+    >
+        <Key
+            android:codes="@integer/key_symbol"
+            android:keyLabel="@string/label_alpha_key"
+            android:keyWidth="20%p"
+            android:isModifier="true"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:codes="@integer/key_f1"
+            android:isModifier="true" />
+        <Key
+            android:codes="@integer/key_space"
+            android:keyIcon="@drawable/sym_keyboard_space"
+            android:iconPreview="@drawable/sym_keyboard_feedback_space"
+            android:keyWidth="40%p"
+            android:isModifier="true" />
+        <Key
+            android:keyLabel="."
+            android:keyIcon="@drawable/hint_popup"
+            android:popupKeyboard="@xml/popup_punctuation"
+            android:isModifier="true" />
+        <Key
+            android:codes="@integer/key_return"
+            android:keyIcon="@drawable/sym_keyboard_return"
+            android:iconPreview="@drawable/sym_keyboard_feedback_return"
+            android:keyWidth="20%p"
+            android:isModifier="true"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row
+        android:keyboardMode="@+id/mode_symbols_with_settings_key"
+        android:rowEdgeFlags="bottom"
+    >
+        <Key
+            android:codes="@integer/key_symbol"
+            android:keyLabel="@string/label_alpha_key"
+            android:keyWidth="15%p"
+            android:isModifier="true"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:codes="@integer/key_settings"
+            android:keyIcon="@drawable/sym_keyboard_settings"
+            android:iconPreview="@drawable/sym_keyboard_feedback_settings"
+            android:isModifier="true" />
+        <Key
+            android:codes="@integer/key_f1"
+            android:isModifier="true" />
+        <Key
+            android:codes="@integer/key_space"
+            android:keyIcon="@drawable/sym_keyboard_space"
+            android:iconPreview="@drawable/sym_keyboard_feedback_space"
+            android:keyWidth="30%p"
+            android:isModifier="true" />
+        <Key
+            android:keyLabel="."
+            android:keyIcon="@drawable/hint_popup"
+            android:popupKeyboard="@xml/popup_punctuation"
+            android:isModifier="true" />
+        <Key
+            android:codes="@integer/key_return"
+            android:keyIcon="@drawable/sym_keyboard_return"
+            android:iconPreview="@drawable/sym_keyboard_feedback_return"
+            android:keyWidth="25%p"
+            android:isModifier="true"
+            android:keyEdgeFlags="right" />
+    </Row>
+</Keyboard>
diff --git a/java/res/xml-fa/kbd_symbols_black.xml b/java/res/xml-fa/kbd_symbols_black.xml
new file mode 100644
index 0000000..37f88ce
--- /dev/null
+++ b/java/res/xml-fa/kbd_symbols_black.xml
@@ -0,0 +1,210 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+/*
+**
+** Copyright 2010, The Android Open Source Project
+**
+** Licensed under the Apache License, Version 2.0 (the "License");
+** you may not use this file except in compliance with the License.
+** You may obtain a copy of the License at
+**
+**     http://www.apache.org/licenses/LICENSE-2.0
+**
+** Unless required by applicable law or agreed to in writing, software
+** distributed under the License is distributed on an "AS IS" BASIS,
+** WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+** See the License for the specific language governing permissions and
+** limitations under the License.
+*/
+-->
+<!-- By Eyad Aboulouz -->
+<Keyboard
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    android:keyWidth="10%p"
+    android:horizontalGap="0px"
+    android:verticalGap="@dimen/key_bottom_gap"
+    android:keyHeight="@dimen/key_height"
+>
+    <Row
+        android:rowEdgeFlags="top"
+    >
+        <Key
+            android:keyLabel="┘ة"
+            android:keyEdgeFlags="left"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="1" />
+        <Key
+            android:keyLabel="┘ت"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="2" />
+        <Key
+            android:keyLabel="┘ث"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="3" />
+        <Key
+            android:keyLabel="█┤"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="4" />
+        <Key
+            android:keyLabel="█╡"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="5" />
+        <Key
+            android:keyLabel="█╢"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="6" />
+        <Key
+            android:keyLabel="┘د"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="7" />
+        <Key
+            android:keyLabel="┘ذ"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="8" />
+        <Key
+            android:keyLabel="┘ر"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="9" />
+        <Key
+            android:keyLabel="┘ب"
+            android:keyEdgeFlags="right"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="0" />
+    </Row>
+    <Row>
+        <Key
+            android:keyLabel="\@"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:keyLabel="\#" />
+        <Key
+            android:keyLabel="$"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="┬ت┬ثظéش┬حظéثظéجظé▒" />
+        <Key
+            android:keyLabel="┘ز"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="%" />
+        <Key
+            android:keyLabel="\u0026" />
+        <Key
+            android:keyLabel="*"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="ظءà┘ص" />
+        <Key
+            android:keyLabel="-"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="_ظôظ¤" />
+        <Key
+            android:keyLabel="+"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="┬▒" />
+        <Key
+            android:keyLabel="("
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="[{\u003C" />
+        <Key
+            android:keyLabel=")"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="]}\u003E"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row>
+        <Key
+            android:codes="@integer/key_shift"
+            android:keyLabel="@string/label_alt_key"
+            android:keyWidth="15%p"
+            android:isModifier="true"
+            android:isSticky="true"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:keyLabel="!"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="┬ة" />
+        <Key
+            android:keyLabel="\u0022"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="ظ£ظإ┬س┬╗╦إ" />
+        <Key
+            android:keyLabel="\'"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="ظءظآ" />
+        <Key
+            android:keyLabel=":" />
+        <Key
+            android:keyLabel="╪ؤ"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters=";" />
+        <Key
+            android:keyLabel="/" />
+        <Key
+            android:keyLabel="╪ا"
+            android:popupKeyboard="@xml/kbd_popup_template"
+            android:popupCharacters="\?┬┐" />
+        <Key
+            android:codes="@integer/key_delete"
+            android:keyIcon="@drawable/sym_bkeyboard_delete"
+            android:iconPreview="@drawable/sym_keyboard_feedback_delete"
+            android:keyWidth="15%p"
+            android:isRepeatable="true"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row
+        android:keyboardMode="@+id/mode_symbols"
+        android:rowEdgeFlags="bottom"
+    >
+        <Key
+            android:codes="@integer/key_symbol"
+            android:keyLabel="@string/label_alpha_key"
+            android:keyWidth="20%p"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:codes="@integer/key_f1" />
+        <Key
+            android:codes="@integer/key_space"
+            android:keyIcon="@drawable/sym_bkeyboard_space"
+            android:iconPreview="@drawable/sym_keyboard_feedback_space"
+            android:keyWidth="40%p" />
+        <Key
+            android:keyLabel="."
+            android:keyIcon="@drawable/hint_popup"
+            android:popupKeyboard="@xml/popup_punctuation" />
+        <Key
+            android:codes="@integer/key_return"
+            android:keyIcon="@drawable/sym_bkeyboard_return"
+            android:iconPreview="@drawable/sym_keyboard_feedback_return"
+            android:keyWidth="20%p"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row
+        android:keyboardMode="@+id/mode_symbols_with_settings_key"
+        android:rowEdgeFlags="bottom"
+    >
+        <Key
+            android:codes="@integer/key_symbol"
+            android:keyLabel="@string/label_alpha_key"
+            android:keyWidth="15%p"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:codes="@integer/key_settings"
+            android:keyIcon="@drawable/sym_bkeyboard_settings"
+            android:iconPreview="@drawable/sym_keyboard_feedback_settings" />
+        <Key
+            android:codes="@integer/key_f1" />
+        <Key
+            android:codes="@integer/key_space"
+            android:keyIcon="@drawable/sym_bkeyboard_space"
+            android:iconPreview="@drawable/sym_keyboard_feedback_space"
+            android:keyWidth="30%p" />
+        <Key
+            android:keyLabel="."
+            android:keyIcon="@drawable/hint_popup"
+            android:popupKeyboard="@xml/popup_punctuation" />
+        <Key
+            android:codes="@integer/key_return"
+            android:keyIcon="@drawable/sym_bkeyboard_return"
+            android:iconPreview="@drawable/sym_keyboard_feedback_return"
+            android:keyWidth="25%p"
+            android:keyEdgeFlags="right" />
+    </Row>
+</Keyboard>
diff --git a/java/res/xml-fa/popup_comma.xml b/java/res/xml-fa/popup_comma.xml
new file mode 100644
index 0000000..b0f1a53
--- /dev/null
+++ b/java/res/xml-fa/popup_comma.xml
@@ -0,0 +1,39 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+/*
+**
+** Copyright 2010, The Android Open Source Project
+**
+** Licensed under the Apache License, Version 2.0 (the "License");
+** you may not use this file except in compliance with the License.
+** You may obtain a copy of the License at
+**
+**     http://www.apache.org/licenses/LICENSE-2.0
+**
+** Unless required by applicable law or agreed to in writing, software
+** distributed under the License is distributed on an "AS IS" BASIS,
+** WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+** See the License for the specific language governing permissions and
+** limitations under the License.
+*/
+-->
+<!-- By Eyad Aboulouz -->
+<Keyboard
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    android:keyWidth="10%p"
+    android:horizontalGap="0px"
+    android:verticalGap="0px"
+    android:keyHeight="@dimen/popup_key_height"
+>
+    <Row
+        android:rowEdgeFlags="top|bottom"
+    >
+        <Key
+            android:codes="@integer/key_settings"
+            android:keyIcon="@drawable/sym_keyboard_settings"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:keyLabel="╪î"
+            android:keyEdgeFlags="right" />
+    </Row>
+</Keyboard>
diff --git a/java/res/xml-fa/popup_punctuation.xml b/java/res/xml-fa/popup_punctuation.xml
new file mode 100644
index 0000000..62684e1
--- /dev/null
+++ b/java/res/xml-fa/popup_punctuation.xml
@@ -0,0 +1,68 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+/*
+**
+** Copyright 2008, The Android Open Source Project
+**
+** Licensed under the Apache License, Version 2.0 (the "License");
+** you may not use this file except in compliance with the License.
+** You may obtain a copy of the License at
+**
+**     http://www.apache.org/licenses/LICENSE-2.0
+**
+** Unless required by applicable law or agreed to in writing, software
+** distributed under the License is distributed on an "AS IS" BASIS,
+** WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+** See the License for the specific language governing permissions and
+** limitations under the License.
+*/
+-->
+<!-- By Eyad Aboulouz -->
+<Keyboard
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    android:keyWidth="10%p"
+    android:horizontalGap="0px"
+    android:verticalGap="0px"
+    android:keyHeight="@dimen/popup_key_height"
+>
+    <Row
+        android:rowEdgeFlags="top"
+    >
+        <Key
+            android:keyLabel=":"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:keyLabel="/" />
+        <Key
+            android:keyLabel="&amp;" />
+        <Key
+            android:keyLabel="(" />
+        <Key
+            android:keyLabel=")" />
+        <Key
+            android:keyLabel="-" />
+        <Key
+            android:keyLabel="+"
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row
+        android:rowEdgeFlags="bottom"
+    >
+        <Key
+            android:keyLabel="╪ؤ"
+            android:keyEdgeFlags="left" />
+        <Key
+            android:keyLabel="\@" />
+        <Key
+            android:keyLabel="\'" />
+        <Key
+            android:keyLabel="&quot;" />
+        <Key
+            android:keyLabel="╪ا" />
+        <Key
+            android:keyLabel="!" />
+        <Key
+            android:keyLabel="╪î"
+            android:keyEdgeFlags="right" />
+    </Row>
+</Keyboard>
diff --git a/java/res/xml-fa/popup_smileys.xml b/java/res/xml-fa/popup_smileys.xml
new file mode 100644
index 0000000..27a9b2f
--- /dev/null
+++ b/java/res/xml-fa/popup_smileys.xml
@@ -0,0 +1,89 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+/*
+**
+** Copyright 2008, The Android Open Source Project
+**
+** Licensed under the Apache License, Version 2.0 (the "License");
+** you may not use this file except in compliance with the License.
+** You may obtain a copy of the License at
+**
+**     http://www.apache.org/licenses/LICENSE-2.0
+**
+** Unless required by applicable law or agreed to in writing, software
+** distributed under the License is distributed on an "AS IS" BASIS,
+** WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+** See the License for the specific language governing permissions and
+** limitations under the License.
+*/
+-->
+<!-- By Eyad Aboulouz -->
+<Keyboard
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    android:keyWidth="15%p"
+    android:horizontalGap="0px"
+    android:verticalGap="0px"
+    android:keyHeight="@dimen/popup_key_height"
+>
+    <Row
+        android:rowEdgeFlags="top"
+    >
+        <Key
+            android:keyLabel="(-:"
+            android:keyOutputText=":-) "
+            android:keyEdgeFlags="left" />
+        <Key
+            android:keyLabel=")-:"
+            android:keyOutputText=":-( " />
+        <Key
+            android:keyLabel="(-;"
+            android:keyOutputText=";-) " />
+        <Key
+            android:keyLabel="╘-:"
+            android:keyOutputText=":-╘ " />
+        <Key
+            android:keyLabel="O-="
+            android:keyOutputText="=-O "
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row>
+        <Key
+            android:keyLabel="*-:"
+            android:keyOutputText=":-* "
+            android:keyEdgeFlags="left" />
+        <Key
+            android:keyLabel="O:"
+            android:keyOutputText=":O " />
+        <Key
+            android:keyLabel="(-B"
+            android:keyOutputText="B-) " />
+        <Key
+            android:keyLabel="$-:"
+            android:keyOutputText=":-$ " />
+        <Key
+            android:keyLabel="3&gt;"
+            android:keyOutputText="&lt;3 "
+            android:keyEdgeFlags="right" />
+    </Row>
+    <Row
+        android:rowEdgeFlags="bottom"
+    >
+        <Key
+            android:keyLabel="]-:"
+            android:keyOutputText=":-[ "
+            android:keyEdgeFlags="left" />
+        <Key
+            android:keyLabel="(-:O"
+            android:keyOutputText="O:-) " />
+        <Key
+            android:keyLabel="\\-:"
+            android:keyOutputText=":-\\ " />
+        <Key
+            android:keyLabel=")':"
+            android:keyOutputText=":'( " />
+        <Key
+            android:keyLabel="D-:"
+            android:keyOutputText=":-D "
+            android:keyEdgeFlags="right" />
+    </Row>
+</Keyboard>
